pkg/crd/v2/multiversion_test.go

```diff
@@ -0,0 +1,165 @@
+package v2
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type multiVersionTestcase struct {
+	inputPackage string
+	types        []string
+
+	listFilesFn listFilesFn
+	listDirsFn  listDirsFn
+	// map of path to file content.
+	inputFiles map[string][]byte
+
+	expectedCrdSpecs map[schema.GroupKind][]byte
+}
+
+func TestMultiVerGenerate(t *testing.T) {
+	testcases := []multiVersionTestcase{
+		{
+			inputPackage: "github.com/myorg/myapi",
+			types:        []string{"Toy"},
+			listDirsFn: func(pkgPath string) (strings []string, e error) {
+				return []string{"v1", "v1alpha1"}, nil
+			},
+			listFilesFn: func(pkgPath string) (s string, strings []string, e error) {
+				return pkgPath, []string{"types.go"}, nil
+			},
+			inputFiles: map[string][]byte{
+				"github.com/myorg/myapi/v1/types.go": []byte(`
+package v1
+
+// +groupName=foo.bar.com
+// +versionName=v1
+
+// +kubebuilder:resource:path=toys,shortName=to;ty
+// +kubebuilder:singular=toy
+
+// Toy is a toy struct
+type Toy struct {
+	// +kubebuilder:validation:Maximum=90
+	// +kubebuilder:validation:Minimum=1
+
+	// Replicas is a number
+	Replicas int32 ` + "`" + `json:"replicas"` + "`" + `
+}
+`),
+				"github.com/myorg/myapi/v1alpha1/types.go": []byte(`
+package v1alpha1
+
+// +groupName=foo.bar.com
+// +versionName=v1alpha1
+
+// +kubebuilder:resource:path=toys,shortName=to;ty
+// +kubebuilder:singular=toy
+
+// Toy is a toy struct
+type Toy struct {
+	// +kubebuilder:validation:MaxLength=15
+	// +kubebuilder:validation:MinLength=1
+
+	// Name is a string
+	Name string ` + "`" + `json:"name,omitempty"` + "`" + `
+
+	// +kubebuilder:validation:Maximum=100
+	// +kubebuilder:validation:Minimum=1
+
+	// Replicas is a number
+	Replicas int32 ` + "`" + `json:"replicas"` + "`" + `
+}
+`),
+			},
+			expectedCrdSpecs: map[schema.GroupKind][]byte{
+				schema.GroupKind{Group: "foo.bar.com", Kind: "Toy"}: []byte(`group: foo.bar.com
+names:
```

Comment:
Just a thought, if we need to write more tests, we can think about having these in their own separate yaml files.

Reply:
Generally, the best practice is that "unit tests should not access filesystem".
Echo some benefits mentioned in https://github.com/spf13/afero#using-afero-for-testing
```
Much faster than performing I/O operations on disk
Avoid security issues and permissions
Far more control. 'rm -rf /' with confidence
Test setup is far more easier to do
No test cleanup needed
```
We can revisit this if it grow to too large.